enigma-runtime-t/src/lib.rs

```diff
@@ -185,6 +210,64 @@ impl Runtime {
         Ok(())
     }
 
+    fn treat_gas_overflow(&mut self, val: &Option<u64>) -> Result<()>{
+        if val.is_none() {
+            self.gas_counter = self.gas_limit;
+            Err(WasmError::GasLimit)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    fn treat_gas_underflow(&mut self, val: &Option<u64>) -> (bool, u64){
+        let mut result = 0;
+        if val.is_none() {
+            self.gas_return = self.gas_limit;
+            (true, result)
+        } else {
+            (false, val.unwrap())
+        }
+    }
+
+    /// args:
+   /// * `new_value_len` - the length of of the new value to be written under the `key`
+   /// * `key` - the key to write the new value
+   /// calculate the gas to be charged for the writing of the new value.
+   /// There is an initial constant value charged for the writing
+   /// If the new value is larger than the old one, then gas is charged for the new bytes.
+   /// If the new value is smaller than the old one, then the gas is returned for the removed bytes.
+    fn calculate_gas_for_writing(&mut self, new_value_len: u64, key: &str) -> Result<u64> {
+        let mut result = Some(0);
+        let val = self.post_execution_state.json[key].clone();
+        let mut old_value_len = 0;
+        // forcing the length of Null value to be 0, since it is not 0.
+        if !val.is_null() {
+            let old_value_vec =
+                serde_json::to_vec(&val).expect("Failed converting Value to vec in Runtime while reading state");
+            old_value_len = old_value_vec.len() as u64;
+        }
+        // If the new value is larger than the old one, the gas should be charged
+        if new_value_len >= old_value_len {
+            let checked_val = (new_value_len - old_value_len).checked_mul(self.gas_costs.write_additional_byte);
+            self.treat_gas_overflow(&checked_val)?;
+            result = self.gas_costs.write_value.checked_add(checked_val.unwrap());
+            self.treat_gas_overflow(&result)?;
+        } // If the new value is smaller than the old one, the gas should be returned
+        else {
+            let decrease_cost = (old_value_len - new_value_len).checked_mul(self.gas_costs.write_additional_byte);
+            let (underflow, val) = self.treat_gas_underflow(&decrease_cost);
```

Comment:
I'm not sure how I understand how is this underflowing. `old_value_len` is bigger than `new_value_len` so subtracting it will yield a positive number, and then you're multiplying.